src/PaginatedData.elm

```diff
@@ -1,353 +1,416 @@
+this function to help determine what pages you ought to request right now.
 
+When you do issue the requests, it is important to call `setPageLoading` here!
+Otherwise, when you call this function again, you'll be told to fetch the page
+again. If you mark the page as `Loading`, then this function won't tell you to
+load it again.
 
-{-| Fetch all existing pages.
+Note that this function does not tell you to re-fetch pages that have failed.
+When to do that is very much dependent on your situation, so you'll need to
+check for that yourself.
 
-Next page is fetched as the previous one arrives successfully.
+This function is a nice helper for the
+[`andThenFetch`](https://package.elm-lang.org/packages/Gizra/elm-essentials/latest/Gizra-Update#andThenFetch)
+function, but is useful in other contexts as well.
+
+If you'd like to fetch all the pages (one at a time), rather than just the
+current page and possibly the next page, take a look at `fetchAllPages` instead.
 
 -}
-fetchAll :
-    ( a, EveryDict a (WebData (PaginatedData key value)) )
-    -> (Int -> c)
-    -> List (Maybe c)
-fetchAll ( backendIndentifier, backendDict ) func =
-    let
-        existingData =
-            EveryDict.get backendIndentifier backendDict
-                |> Maybe.withDefault RemoteData.NotAsked
+fetchNextPage : Int -> PaginatedData e k v -> List Int
+fetchNextPage currentPage (PaginatedData { pager }) =
+    if currentPage < 1 then
+        []
+
+    else
+        let
+            currentPageData =
+                Dict.get currentPage pager
+                    |> Maybe.withDefault NotAsked
+        in
+        case currentPageData of
+            NotAsked ->
+                -- If we haven't asked for the current page yet, we should do so.
+                [ currentPage ]
+
+            Loading ->
+                -- If the current page is loading, we wait for an answer.
+                []
+
+            Failure _ ->
+                -- We don't automatically re-try failures ... and, if we got a
+                -- failure, we don't try to pre-fetch the next page.
+                []
+
+            Success _ ->
+                -- If we have data for the current page, we check whether to fetch
+                -- the next page.
+                let
+                    nextPage =
+                        currentPage + 1
+                in
+                case Dict.get nextPage pager of
+                    Just NotAsked ->
+                        -- If we expect a page to exist here, and it is a `NotAsked`,
+                        -- then we ask for it.
+                        [ nextPage ]
+
+                    Just _ ->
+                        -- If we expect a page to exist here, and it's something other
+                        -- than `NotAsked`, we don't do anything.
+                        []
+
+                    Nothing ->
+                        -- If we don't expect a page to exist here, and this is not
```

Comment:
I'm not sure I follow what you mean here?

Reply:
By "expect a page to be here" I mean: do we expect this page to exist? We create `NotAsked` entries for every page we expect to exist when we process `handleFetchedPage`, based on the `totalCount` and what we infer the page size to be. So, if we get a `Nothing` here, it means that we don't expect the next page to exist. (Or, to put it another, we think the current page is the last page).

Reply:
> So, if we get a Nothing here, it means that we don't expect the next page to exist. (Or, to put it another, we think the current page is the last page).

👍 I think it's worth adding ^^ in the comment